adapters/mqtt-vertx-base/src/main/java/org/eclipse/hono/adapter/mqtt/AbstractVertxBasedMqttProtocolAdapter.java

```diff
@@ -797,4 +889,56 @@ protected void onMessageSent(final MqttContext ctx) {
      */
     protected void onMessageUndeliverable(final MqttContext ctx) {
     }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    protected void onCommandReceived(final String tenantId, final String deviceId, final ProtonDelivery delivery, final Message message) {
+        super.onCommandReceived(tenantId, deviceId, delivery, message);
+
+        final MqttEndpoint endpoint = endpoints.get(Device.asAddress(tenantId, deviceId));
+        if(endpoint!=null) {
+            final String cmdReqId = validateAndGenerateCommandRequestId(tenantId,deviceId,message).get(); // TODO: why an optional?
+            endpoint.publish(String.format("%s/%s/%s", COMMAND_REQUEST_PREFIX, message.getSubject(),cmdReqId),
+                    MessageHelper.getPayload(message), MqttQoS.AT_LEAST_ONCE, false, false);
+        }
+        else {
+            LOG.error(
+                    "received command could not be delivered since no endpoint for device: [{}] from tenant: [{}] could be found",
+                    deviceId, tenantId);
+        }
+    }
+
+    /**
+     * Sends the event of a subscribed device, that is now online (ttd==-1), or disconnects (ttd=0).
+     *
+     * @param authenticatedDevice The device, that wants to handle commands.
+     * @param ttd The time til disconnect.
+     */
+    protected void sendTtdEvent(final Device authenticatedDevice, final Integer ttd) {
```

Comment:
IMHO we have retrieved the TenantObject, the registrationAssertion and the event sender already at this point. Why do that again? Better would be to pass them in?

Reply:
I think the check for the tenant could be removed. The assertion comes from cache, so this should not be a problem. I do not think, that the sender exist in any case?

Reply:
The `TenantObject` comes from the cache as well. So basically we could remove the tenant, but:
I thought we want to make the behaviour if the MQTT adapter auto-sends a `ttd` event configurable per tenant in the future, and then the TenantObject would be needed here again.

The sender indeed is probably not present at this point of time (when the device just connected), so this cannot be eliminated, I agree.

But I have another proposal: 
very recently @sophokles73 introduced the `MqttContext` class, which could be used here to register key-value pairs.
For the authenticated device there is already a constructor available, but the generic `put` method is good for any pairs.
So why not put the TenantObject and the assertionToken into the context (which currently is constructed at the very end of `onAuthenticationSuccess` and pass that around?
This would IMHO solve the problem in an elegant way without repeating code...
WDYT?

Reply:
That might be a good solution but I think it is an optimization, that should be done in a second step.

Reply:
Sounds reasonable :+1: . We'll do this in another step then.
IMHO the code can remain unchanged for the first step then.